### ISSUE TITLE
Weird error in 1.10.0 trying to parse site-packages\torch\_C\_VariableFunctions.pyi

### ISSUE THREAD

**Issue by lovettchris** [python/mypy#17189]
**Bug Report**

Unexpected error messages from mypy about pytorch "C" wrappers containing non-utf-8 chars.

**To Reproduce**

Setup python 3.10 environment with the following installed:
```
pytorch-lightning         2.2.3
torch                     2.3.0+cu118
torchaudio                2.3.0+cu118
torchmetrics              1.3.2
torchvision               0.18.0+cu118
```

```python
import torch
print(torch.Tensor([1,2,3]))
```

Run `mypy . --check`

**Expected Behavior**

Should not produce an error.

**Actual Behavior**

mypy: can't decode file 'd:\Anaconda3\envs\sr\lib\site-packages\torch\_C\_VariableFunctions.pyi': 'utf-8' codec can't decode byte 0x93 in position 290995: invalid start byte

**Your Environment**

- Mypy version used: 1.10.0
- Mypy command-line flags: --check
- Mypy configuration options from `mypy.ini` (and other config files): 
- Python version used: 3.10
- - OS: Windows 11 23H2

This error is new to 1.10.0, did not happen in 1.9.0.

**Comment by PaulFidika** [comment]
I have this same issues; it doesn't like a file in the pytorch for some reason.

**Comment by JelleZijlstra** [comment]
It would be helpful in someone can provide a reproduction case that doesn't involve installing all of pytorch.

**Comment by lovettchris** [comment]
And I've attached the file in case you want to check it:
[_VariableFunctions.zip](https://github.com/python/mypy/files/15179697/_VariableFunctions.zip)

Looks like it is generated from [_VariableFunctions.pyi.in](https://github.com/pytorch/pytorch/blob/main/torch/_C/_VariableFunctions.pyi.in)

And the offset reported by mypy corresponds to line 6320 column 84 which contains these curly quotes:

![image](https://github.com/python/mypy/assets/18707114/36d43ada-a6da-4775-8609-a2773578be0d)

Normally these would need to be utf-8 encoded, but in this case they are not and they are stored in the file as byte value 147 which is indeed an invalid utf-8 encoding, so we should file a bug against pytorch.  When encoded as utf-8 the curly braces become a 3 byte sequence 239, 191, 189.  But it is a bit complicated figuring out where this is coming from - it appears this _variablefunctions template is expanding a bunch of numpy functions, this one is the [cov function](https://numpy.org/doc/stable/reference/generated/numpy.cov.html).  That function is in my env at D:\Anaconda3\envs\sr\Lib\site-packages\numpy\lib\function_base.py but the line containing that comment does not have curly quotes.

Mypy should add line/col and all this information so bugs in packages can be properly identified and reported.  Or mypy should forget it and be more forgiving about bad utf-8 encodings.

**Comment by JelleZijlstra** [comment]
Thanks. I'm not sure this is something mypy can realistically fix: if we can't decode the file as UTF-8, we can't even know where the line boundaries are, so we can hardly report a line number.

Odd though that this error is apparently new in 1.10. Perhaps in 1.9 we didn't attempt to parse this file for whatever reason.

**Comment by lovettchris** [comment]
You can identify newlines in utf-8 encoded bytes.  See https://en.wikipedia.org/wiki/UTF-8.  The second to fourth bytes of a utf-8 sequence can never be '\n' because they always have the high bit set.  Column position is more difficult, but identifying the line number would have been helpful in this case.

![image](https://github.com/python/mypy/assets/18707114/6241974b-8782-4083-9243-8029abf3e7fe)

**Comment by JelleZijlstra** [comment]
But the file we're discussing here is not in UTF-8. I tried it with all the encodings that Python comes with, and the relevant line can be decoded by many of them:

<details>

```

            These relative weights are typically large for observations considered ìimportantî and smaller for
mac_romanian
            These relative weights are typically large for observations considered …importantî and smaller for
mac_farsi
èÇÁËÁÊÁ%/ÈÑÎÁÏÁÑÅÇÈË/ÊÁÈ`øÑ[/%%`%/ÊÅÁÃ?Ê?ÂËÁÊÎ/ÈÑ?>Ë[?>ËÑÀÁÊÁÀlÑ_ø?ÊÈ/>Èm/>ÀË_/%%ÁÊÃ?Ê
cp273
            These relative weights are typically large for observations considered important and smaller for
raw_unicode_escape
            These relative weights are typically large for observations considered “important” and smaller for
cp1252
            These relative weights are typically large for observations considered important and smaller for
iso8859_14
            These relative weights are typically large for observations considered important and smaller for
iso8859_2
            These relative weights are typically large for observations considered …importantî and smaller for
mac_arabic
            These relative weights are typically large for observations considered ìimportantî and smaller for
mac_croatian
            These relative weights are typically large for observations considered “important” and smaller for
cp1256
            These relative weights are typically large for observations considered important and smaller for
iso8859_6
            These relative weights are typically large for observations considered important and smaller for
iso8859_10
èÇÁËÁÊÁ%/ÈÑÎÁÏÁÑÅÇÈË/ÊÁÈ`øÑÄ/%%`%/ÊÅÁÃ?Ê?ÂËÁÊÎ/ÈÑ?>ËÄ?>ËÑÀÁÊÁÀlÑ_ø?ÊÈ/>Èm/>ÀË_/%%ÁÊÃ?Ê
cp1140
            These relative weights are typically large for observations considered УimportantФ and smaller for
cp1125
            These relative weights are typically large for observations considered “important” and smaller for
cp1257
            These relative weights are typically large for observations considered ôimportantö and smaller for
cp858
            These relative weights are typically large for observations considered important and smaller for
iso8859_7
            These relative weights are typically large for observations considered important and smaller for
iso8859_11
            These relative weights are typically large for observations considered important and smaller for
hp_roman8
            These relative weights are typically large for observations considered ⌠important■ and smaller for
koi8_r
            These relative weights are typically large for observations considered “important” and smaller for
cp1253
            These relative weights are typically large for observations considered important and smaller for
iso8859_15
            These relative weights are typically large for observations considered ìimportantî and smaller for
mac_iceland
            These relative weights are typically large for observations considered important and smaller for
iso8859_3
            These relative weights are typically large for observations considered ™importantî and smaller for
mac_greek
††††††周敳攠牥污瑩癥⁷敩杨瑳⁡牥⁴祰楣慬汹⁬慲来⁦潲⁯扳敲癡瑩潮猠捯湳楤敲敤ₓ業灯牴慮璔⁡湤⁳浡汬敲⁦潲
utf_16_be
            These relative weights are typically large for observations considered ôimportantË and smaller for
cp863
            These relative weights are typically large for observations considered important and smaller for
iso8859_8
            These relative weights are typically large for observations considered ôimportantö and smaller for
cp857
            These relative weights are typically large for observations considered “important” and smaller for
cp1258
            These relative weights are typically large for observations considered ďimportantĒ and smaller for
mac_latin2
            These relative weights are typically large for observations considered ōimportantö and smaller for
cp775
            These relative weights are typically large for observations considered ìimportantî and smaller for
mac_roman
            These relative weights are typically large for observations considered ôimportantö and smaller for
cp852
            These relative weights are typically large for observations considered УimportantФ and smaller for
cp866
            These relative weights are typically large for observations considered ôimportant¤ and smaller for
cp720
            These relative weights are typically large for observations considered ףimportantפ and smaller for
cp862
            These relative weights are typically large for observations considered ôimportantö and smaller for
cp437
            These relative weights are typically large for observations considered “important” and smaller for
palmos
            These relative weights are typically large for observations considered important and smaller for
iso8859_9
            These relative weights are typically large for observations considered ףimportantפ and smaller for
cp856
            These relative weights are typically large for observations considered important and smaller for
latin_1
ΝΪΧΉΧΈΧ%/ ΫΌΧΎΧΫΩΪ Ή/ΈΧ `¨ΫΥ/%%`%/ΈΩΧΨ?Έ?ΤΉΧΈΌ/ Ϋ?>ΉΥ?>ΉΫΦΧΈΧΦlΫ_¨?Έ /> m/>ΦΉ_/%%ΧΈΨ?Έ
cp875
            These relative weights are typically large for observations considered important and smaller for
unicode_escape
            These relative weights are typically large for observations considered ΥimportantΦ and smaller for
cp737
            These relative weights are typically large for observations considered ôimportantö and smaller for
cp865
            These relative weights are typically large for observations considered “important” and smaller for
ptcp154
            These relative weights are typically large for observations considered ôimportantö and smaller for
cp861
            These relative weights are typically large for observations considered Њimportantћ and smaller for
cp855
èÇÁËÁÊÁ%/ÈÑÎÁÏÁÑÅÇÈË/ÊÁÈ`øÑÄ/%%`%/ÊÅÁÃ?Ê?ÂËÁÊÎ/ÈÑ?>ËÄ?>ËÑÀÁÊÁÀlÑ_ø?ÊÈ/>Èm/>ÀË_/%%ÁÊÃ?Ê
cp500
            These relative weights are typically large for observations considered ôimportantõ and smaller for
cp860
††††††桔獥⁥敲慬楴敶眠楥桧獴愠敲琠灹捩污祬氠牡敧映牯漠獢牥慶楴湯⁳潣獮摩牥摥錠浩潰瑲湡鑴愠摮猠慭汬牥映牯
utf_16_le
            These relative weights are typically large for observations considered “important” and smaller for
cp874
            These relative weights are typically large for observations considered ôimportantö and smaller for
cp850
            These relative weights are typically large for observations considered ±important½ and smaller for
cp864
            These relative weights are typically large for observations considered ⌠important■ and smaller for
koi8_u
            These relative weights are typically large for observations considered “important” and smaller for
cp1254
††††††桔獥⁥敲慬楴敶眠楥桧獴愠敲琠灹捩污祬氠牡敧映牯漠獢牥慶楴湯⁳潣獮摩牥摥錠浩潰瑲湡鑴愠摮猠慭汬牥映牯
utf_16
            These relative weights are typically large for observations considered important and smaller for
iso8859_4
            These relative weights are typically large for observations considered “important” and smaller for
cp1250
            These relative weights are typically large for observations considered important and smaller for
iso8859_16
            These relative weights are typically large for observations considered УimportantФ and smaller for
mac_cyrillic
            These relative weights are typically large for observations considered important and smaller for
tis_620
èÇÁËÁÊÁ%/ÈÑÎÁÏÁÑÅÇÈË/ÊÁÈ`øÑÄ/%%`%/ÊÅÁÃ?Ê?ÂËÁÊÎ/ÈÑ?>ËÄ?>ËÑÀÁÊÁÀlÑ_ø?ÊÈ/>Èm/>ÀË_/%%ÁÊÃ?Ê
cp037
            These relative weights are typically large for observations considered important and smaller for
cp1006
            These relative weights are typically large for observations considered “important” and smaller for
cp1251
            These relative weights are typically large for observations considered ìimportantî and smaller for
mac_turkish
            These relative weights are typically large for observations considered important and smaller for
iso8859_1
            These relative weights are typically large for observations considered “important” and smaller for
kz1048
            These relative weights are typically large for observations considered “important” and smaller for
koi8_t
            These relative weights are typically large for observations considered “important” and smaller for
cp1255
è[ÁËÁÊÁ%/ÈÑÎÁÏÁÑÅ[ÈË/ÊÁÈıøÑÄ/%%ı%/ÊÅÁÃ?Ê?ÂËÁÊÎ/ÈÑ?>ËÄ?>ËÑÀÁÊÁÀlÑ_ø?ÊÈ/>Èm/>ÀË_/%%ÁÊÃ?Ê
cp1026
            These relative weights are typically large for observations considered important and smaller for
charmap
            These relative weights are typically large for observations considered important and smaller for
iso8859_5
            These relative weights are typically large for observations considered important and smaller for
iso8859_13

```

</details>

Now, in this case it's pretty clear to a human the file is in some "extended ASCII" encoding like cp1252, but in general that is not a safe assumption; it might be in some totally different encoding that uses a different byte sequence for newlines. I don't think mypy should get into the business of guessing the encoding if the default encoding is "nearly" right.

I'm closing this as there isn't a mypy bug: the file is not valid UTF-8, and mypy correctly reported as much.

**Comment by lovettchris** [comment]
I believe the default encoding for python is utf-8 these days isn't it ?
https://peps.python.org/pep-0686/

**Comment by ethanhs** [comment]
The default will only change in 3.15 and above, and even then files created by text editors may not be utf-8.
